core/src/main/java/org/infinispan/statetransfer/StateConsumerImpl.java

```diff
@@ -557,135 +566,176 @@ public void applyState(final Address sender, int topologyId, boolean pushTransfe
-               transactionManager.begin();
-               ctx = icf.createInvocationContext(transactionManager.getTransaction(), true);
-               ((TxInvocationContext) ctx).getCacheTransaction().setStateTransferFlag(PUT_FOR_STATE_TRANSFER);
+            localTransaction.setStateTransferFlag(PUT_FOR_STATE_TRANSFER);
+            for (InternalCacheEntry e : cacheEntries) {
+               // CallInterceptor will preserve the timestamps if the metadata is an InternalMetadataImpl instance
+               key = e.getKey();
+               CompletableFuture<?> future = invokePut(segmentId, ctx, e);
+               if (!future.isDone()) {
+                  throw new IllegalStateException("State transfer in-tx put should always be synchronous");
+               }
+            }
+         } catch (Throwable t) {
+            logApplyException(t, key);
+            if (localTransaction != null) {
+               return invokeRollback(localTransaction).handle((rv, t1) -> {
+                  transactionTable.removeLocalTransaction(localTransaction);
+                  if (t1 != null) {
+                     t.addSuppressed(t1);
+                  }
+                  return null;
+               });
             } else {
-               // non-tx cache
-               ctx = icf.createSingleKeyNonTxInvocationContext();
+               return CompletableFutures.completedNull();
             }
+         }
 
-            // CallInterceptor will preserve the timestamps if the metadata is an InternalMetadataImpl instance
-            InternalMetadataImpl metadata = new InternalMetadataImpl(e);
-            PutKeyValueCommand put = commandsFactory.buildPutKeyValueCommand(e.getKey(), e.getValue(), segmentId,
-                                                                             metadata, STATE_TRANSFER_FLAGS);
-            ctx.setLockOwner(put.getKeyLockOwner());
-            interceptorChain.invoke(ctx, put);
-
-            if (transactionManager != null) {
-               transactionManager.commit();
-            }
-         } catch (Exception ex) {
-            if (!cache.wired().getStatus().allowInvocations()) {
-               log.debugf("Cache %s is shutting down, stopping state transfer", cacheName);
-               break;
-            } else {
-               log.problemApplyingStateForKey(ex.getMessage(), e.getKey(), ex);
+         return invoke1PCPrepare(localTransaction).whenComplete((rv, t) -> {
+            transactionTable.removeLocalTransaction(localTransaction);
+            if (t != null) {
+               logApplyException(t, NO_KEY);
             }
-         } finally {
-            try {
-               if (transactional && transactionManager.getTransaction() != null) {
-                  transactionManager.rollback();
+         });
+      } else {
+         // non-tx cache
+         AggregateCompletionStage<Void> aggregateStage = CompletionStages.aggregateCompletionStage();
+         for (InternalCacheEntry e : cacheEntries) {
```

Comment:
Should we be using `putAll` here instead? This would only require a single context and would allow for batching with stores.

Reply:
I also thought about using `putAll`, but it would require extending `PutMapCommand` to support different metadata instances for each entry and it seemed too big of a change.

Reply:
Gotcha, makes sense.